lib/xrp/XrpRpc.js

```diff
@@ -0,0 +1,237 @@
+    let prepared = await this.rpc.preparePayment(sourceAddress, payment);
+    await this.rpc.disconnect();
+    return prepared;
+  }
+
+  async signTransaction({ address, amount, passphrase, tag, invoiceID, secret }) {
+    if (passphrase && !secret) {
+      secret = passphrase;
+    }
+    if (!secret) {
+      throw new Error('Secret not provided');
+    }
+    let sourceAddress = this.rpc.deriveAddress(this.rpc.deriveKeypair(secret).publicKey);
+    const payment = {
+      source: {
+        address: sourceAddress,
+        tag: tag || undefined,
+        maxAmount: {
+          value: amount.toString(),
+          currency: 'XRP'
+        }
+      },
+      destination: {
+        address: address,
+        tag: tag || undefined,
+        amount: {
+          value: amount.toString(),
+          currency: 'XRP'
+        }
+      },
+      invoiceID: invoiceID || undefined,
+    };
+    return this.rpc.preparePayment(sourceAddress, payment).then(async (prepared) => {
+      const { signedTransaction } = this.rpc.sign(prepared.txJSON, secret);
+      return signedTransaction;
+    });
+  }
+
+  async submitSignedTransaction({ signedTx }, alreadyConnected = false) {
+    if (!alreadyConnected) {
+      await this.rpc.connect();
+    }
+    return this.rpc.submit(signedTx).then(async (response) => {
+      if (!alreadyConnected) {
+        await this.rpc.disconnect();
+      }
+      if (response.resultCode !== 'tesSUCCESS') {
+        return new Error('Failed to submit transaction.');
+      }
+      return response.tx_json.hash;
+    });
+  }
+
+  async submitSignedTransactionMany({ signedTxArray }) {
+    let resultArray = [];
+    for (const signedTx of signedTxArray) {
+      const emitData = { signedTx };
+      try {
+        let txHash = await this.submitSignedTransaction({ signedTx });
+        emitData.txid = txHash;
+        resultArray.push(emitData);
+        this.emitter.emit('success', emitData);
+      } catch (e) {
+        emitData.error = e;
+        resultArray.push(emitData);
+        this.emitter.emit('success', emitData);
+      }
+    }
+    return resultArray;
+  }
+
+  async unlockAndSendToAddressMany({ payToArray, passphrase }) {
+    if (passphrase === undefined) {
+      passphrase = await passwordPromptAsync('> ');
+    }
+
+    const resultArray = [];
+    await this.rpc.connect();
+    for (const payment of payToArray) {
+      const { address, amount, id } = payment;
+      const emitData = { address, amount, id };
+      this.emitter.emit('attempt', emitData);
+      try {
+        const txid = await this.sendToAddress({ address, amount, passphrase }, true);
+        emitData.txid = txid;
+        resultArray.push(emitData);
+        this.emitter.emit('success', emitData);
+
+        //do not await confirmations,
+      } catch(e) {
+        emitData.error = e;
+        resultArray.push(emitData);
+        this.emitter.emit('failure', emitData);
+      }
+    }
+    return resultArray;
```

Comment:
I think we should disconnect before returning.

Reply:
Fixed